s3/service.go

```diff
@@ -0,0 +1,257 @@
+package s3util
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+)
+
+const DefaultChunkSizeLimit = 1024 * 1024 * 50
+const DefaultIncompleteExpiry = time.Hour * 24
+
+type Service struct {
+	Bucket      string
+	Region      string
+	Timeout     time.Duration
+	Concurrency int // Concurrent downloading
+
+	ChunkSizeLimit         int
+	MinUploadSizeChunked   int
+	IncompleteUploadExpiry time.Duration
+
+	incomplete map[string]*Upload
+	lock       sync.RWMutex
+
+	Session *session.Session
+	Client  *s3.S3
+
+	uploads  chan *Upload
+	shutdown chan bool
+}
+
+func (s *Service) Init(id, secret string) error {
+	sess, err := session.NewSession(&aws.Config{
+		Credentials: credentials.NewStaticCredentials(id, secret, ""),
+		Region:      aws.String(s.Region),
+	})
+	if err != nil {
+		return err
+	}
+
+	if s.ChunkSizeLimit == 0 {
+		s.ChunkSizeLimit = DefaultChunkSizeLimit
+	}
+	if s.MinUploadSizeChunked == 0 {
+		s.MinUploadSizeChunked = DefaultChunkSizeLimit
+	}
+	if s.IncompleteUploadExpiry == 0 {
+		s.IncompleteUploadExpiry = DefaultIncompleteExpiry
+	}
+
+	s.Client = s3.New(sess)
+	s.Session = sess
+	s.uploads = make(chan *Upload, 10)
+	s.shutdown = make(chan bool)
+	s.incomplete = make(map[string]*Upload)
+	return nil
+}
+
+// Checks if the file exists
+func (s *Service) Exists(file string) (bool, error) {
+	_, err := s.Client.HeadObject(&s3.HeadObjectInput{
+		Bucket: aws.String(s.Bucket),
+		Key:    aws.String(file),
+	})
+	if err == nil {
+		return true, nil
+	}
+
+	aerr, ok := err.(awserr.Error)
+	if !ok || aerr.Code() != "NotFound" {
+		return false, err
+	}
+	return false, nil
+}
+
+// Gets a file reader from Service
+func (s *Service) Get(file string) (io.ReadCloser, error) {
+	out, err := s.Client.GetObject(&s3.GetObjectInput{
+		Bucket: aws.String(s.Bucket),
+		Key:    aws.String(file),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return out.Body, nil
+}
+
+func (s *Service) Download(file string, w io.WriterAt) error {
```

Comment:
You make use of the convention "filename" later down in this file. Please stick to one convention. I'd also advise using something else that indicates it's a full file path. I can't recall, but IMO filename indicates a basename not a full conforming URI

Reply:
This applies to "Get" method too.